File: DSA Practice/LeetCode/1161. Maximum Level Sum of a Binary Tree/solution.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun maxLevelSum(root: TreeNode?): Int {

        var levelSumMap : MutableMap<Int, Int> = mutableMapOf<Int, Int>()

        fun rec(node : TreeNode?, level : Int){

            if (level in levelSumMap.keys){
                levelSumMap.put(level, levelSumMap.get(level)!!.plus(node!!.`val`))
            } else {
                levelSumMap.put(level, node!!.`val`)
            }

            if (node!!.left == null && node!!.right == null){
                return
            } else {
                if (node!!.left != null){
                    rec(node!!.left, level.plus(1))
                }

                if (node!!.right != null){
                    rec(node!!.right, level.plus(1))
                }
            }
        }

        var m = -1000000
        var res = 0

        rec(root, 1)

        for (e in levelSumMap){
            if (e.value > m){
                m = e.value
                res = e.key
            }
        }

        return res
        
    }
}

File: DSA Practice/LeetCode/1161. Maximum Level Sum of a Binary Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        levelSumMap = {}

        def rec(node, level):
            if level in levelSumMap.keys():
                levelSumMap[level] += node.val
            else:
                levelSumMap[level] = node.val

            if node.left == None and node.right == None:
                return
            else:
                if node.left != None:
                    rec(node.left, level+1)

                if node.right != None:
                    rec(node.right, level+1)

        rec(root, 1)
        #print(levelSumMap)

        m = -1000000
        res = 0
        for k,v in levelSumMap.items():
            if v > m:
                m = v
                res = k

        return res